homeassistant/components/sensibo/climate.py

```diff
@@ -276,3 +288,43 @@ async def async_assume_state(self, state: str) -> None:
         """Sync state with api."""
         await self._async_set_ac_state_property("on", state != HVACMode.OFF, True)
         await self.coordinator.async_refresh()
+
+    async def async_set_timer(self, state: str, minutes: int | None = None) -> None:
+        """Set or delete timer."""
+        if state == "off" and self.device_data.timer_id is None:
+            raise HomeAssistantError("No timer to delete")
+
+        if state == "on" and minutes is None:
+            raise HomeAssistantError("No value provided for timer")
+
+        timer_on: bool | None = None
+        timer_id: str | None = None
+        timer_state_on: bool | None = None
+        timer_time: datetime | None = None
+
+        if state == "off":
+            result = await self.async_send_command("del_timer")
+        else:
+            new_state = bool(self.device_data.ac_states["on"] is False)
+            params = {
+                "minutesFromNow": minutes,
+                "acState": {**self.device_data.ac_states, "on": new_state},
+            }
+            result = await self.async_send_command("set_timer", params)
+
+        if result["status"] == "success":
+            if state == "on":
+                timer_on = True
+                timer_id = result["result"]["id"]
+                timer_state_on = new_state
+                if TYPE_CHECKING:
+                    assert minutes is not None
+                timer_time = dt.now() + timedelta(minutes=minutes)
+                setattr(self.device_data, "timer_on", timer_on)
+            setattr(self.device_data, "timer_id", timer_id)
+            setattr(self.device_data, "timer_state_on", timer_state_on)
+            setattr(self.device_data, "timer_time", timer_time)
```

Comment:
It would be a lot cleaner if you could do something like
`self.device_date.update_from_command(result)` as it would avoid all these setattr calls

Reply:
I think also it would probably look nicer to have an update function as these setattr calls are a bit everywhere.
But I don't see a better way to update the dataclass except using setattr or maybe use of a function was the point?

Reply:
The dataclass isn't frozen so you should be able to change it directly.

```diff
diff --git a/homeassistant/components/sensibo/climate.py b/homeassistant/components/sensibo/climate.py
index d1045cab96..dda1e9696a 100644
--- a/homeassistant/components/sensibo/climate.py
+++ b/homeassistant/components/sensibo/climate.py
@@ -320,10 +320,10 @@ class SensiboClimate(SensiboDeviceBaseEntity, ClimateEntity):
                 if TYPE_CHECKING:
                     assert minutes is not None
                 timer_time = dt.now() + timedelta(minutes=minutes)
-                setattr(self.device_data, "timer_on", timer_on)
-            setattr(self.device_data, "timer_id", timer_id)
-            setattr(self.device_data, "timer_state_on", timer_state_on)
-            setattr(self.device_data, "timer_time", timer_time)
+                self.device_data.timer_on = timer_on
+            self.device_data.timer_id = timer_id
+            self.device_data.timer_state_on = timer_state_on
+            self.device_data.timer_time = timer_time
             self.async_write_ha_state()
             return
 
```

I'm suggesting binding a method to the dataclass to process and update data from an api response so Home Assistant knows less about the details

Reply:
The response from the api only contains more or less that it was successful and not anything useful to update the data.
So either one has to update manually (like your suggestion) or simply stick a coordinator refresh in instead to update all data.

Response from api on setting a timer:
```json
{"status": "success", "result": {"id": "FLyiYoFYp9"}}
```

Reply:
Could as follow-up PR move the logic of such to the upstream package to return a proper response with the required information and thereby require less knowledge of the data in the HA integration.

Reply:
At the end it didn't matter I noticed since the service is called from `climate.py` but the entities are in `sensor` and `binary_sensor` which doesn't update anyway and didn't immediately see a good way to tell them to refresh so I set to ask the coordinator to update instead which should be pretty near to immediate.
Would be interesting to know if there is a good way to ask all entities to update from one.